src/package-resolver.js

```diff
@@ -530,10 +530,48 @@ export default class PackageResolver {
       this.resolveToResolution(req);
     }
 
+    deps.forEach(dep => {
+      const name = normalizePattern(dep.pattern).name;
+      this.optimizeResolutions(name);
+    });
+
     activity.end();
     this.activity = null;
   }
 
+  // for a given package, see if a single manifest can satisfy all ranges
+  optimizeResolutions(name: string) {
+    const patterns: Array<string> = this.patternsByPackage[name] || [];
+
+    // don't optimize things that already have a lockfile entry:
+    // https://github.com/yarnpkg/yarn/issues/79
+    const anyLocked = patterns.some(pattern => this.lockfile.getLocked(pattern));
+    if (anyLocked) {
+      return;
+    }
+
+    const manifestByVersion: Map<string, Manifest> = new Map();
+    patterns.forEach(p => {
+      const manifest = this.patterns[p];
+      manifestByVersion.set(manifest.version, manifest);
+    });
+
+    // reverse sort, so we'll find the maximum satisfying version
+    const availableVersions = [...manifestByVersion.keys()];
```

Comment:
Why not `Array.from(manifestByVersion.keys())`. That said spread looks is faster in latest V8 versions so NVM: https://jsperf.com/set-iterator-vs-foreach/4